File: testdata/fixtures/elixir/CWE-94.ex
# CWE-94: Code Injection – Code.eval_string with user input executes code.

defmodule Fixture.CWE94 do
  def vuln(user_input) do
    # ruleid: elixir-code-injection
    Code.eval_string(user_input)
    # ruleid: elixir-code-injection
    Code.eval_string(user_input <> "; extra")
    # ruleid: elixir-code-injection
    Code.eval_string(user_input)
    # ruleid: elixir-code-injection
    s = user_input; Code.eval_string(s)
  end

  def safe do
    # ok: elixir-code-injection
    Code.eval_string("1 + 2")
    # ok: elixir-code-injection
    Code.eval_string("fixed")
    # ok: elixir-code-injection
    Code.eval_string("hello")
  end
end

File: testdata/fixtures/ocaml/CWE-94.ml
(* CWE-94: Code Injection – dynamic loading/eval with user input.
 * Dynlink.loadfile and Toploop.execute_phrase with variable are vulnerable.
 *)

let vuln user_input =
  (* ruleid: ocamllint-code-injection *)
  let _ = Dynlink.loadfile user_input in
  (* ruleid: ocamllint-code-injection *)
  let _ = Toploop.execute_phrase user_input in
  (* ruleid: ocamllint-code-injection *)
  let p = user_input in let _ = Dynlink.loadfile p in
  (* ruleid: ocamllint-code-injection *)
  let _ = Toploop.execute_phrase (user_input ^ ";;") in
  ()

let safe () =
  (* ok: ocamllint-code-injection *)
  let _ = Dynlink.loadfile "/tmp/fixed.cma" in
  (* ok: ocamllint-code-injection *)
  let _ = Toploop.execute_phrase "let x = 1;;" in
  (* ok: ocamllint-code-injection *)
  let _ = Dynlink.loadfile "/usr/lib/fixed.cma" in
  ()
